asf_heat_pump_suitability/utils/geo_utils.py

```diff
@@ -88,3 +89,82 @@ def ensure_crs_match(
     if gdf_1.crs != gdf_2.crs:
         gdf_2 = gdf_2.to_crs(gdf_1.crs)
     return gdf_2
+
+
+def load_and_filter_lsoa_geometries(
+    la_lsoas: List[str], lsoa_shp_path: str, target_crs: str
+) -> gpd.GeoDataFrame:
+    """
+    Loads and preprocesses LSOA geospatial data by:
+      - Reading a shapefile.
+      - Converting to a target CRS if necessary.
+      - Filtering LSOAs to just those that belong to the specified local authority.
+
+    Args:
+        la_lsoas (List[str]): List of LSOA codes for the current local authority.
+        lsoa_shp_path (str): Path (local or S3) to the national LSOA shapefile.
+        target_crs (str): The target Coordinate Reference System (e.g., "EPSG:27700").
+
+    Returns:
+        gpd.GeoDataFrame: GeoDataFrame with filtered LSOA geometries in the target CRS.
+
+    Raises:
+        ValueError: If no matching LSOAs are found in the shapefile.
+    """
+    logging.info("Preprocessing LSOA geometries...")
+    lsoa_gdf = gpd.read_file(lsoa_shp_path)
+
+    # Reproject if needed
+    if lsoa_gdf.crs != target_crs:
+        logging.info(f"Converting shapefile to {target_crs}")
+        lsoa_gdf = lsoa_gdf.to_crs(target_crs)
+
+    la_lsoa_geometries_gdf = lsoa_gdf[lsoa_gdf["LSOA21CD"].isin(la_lsoas)]
+    if la_lsoa_geometries_gdf.empty:
+        raise ValueError(
+            "No LSOAs found for this local authority in the provided LSOA shapefile."
+        )
+
+    return la_lsoa_geometries_gdf
+
+
+def merge_hp_suitability_data_with_geometries(
+    hp_suitability_scores_pd: pd.DataFrame,
+    la_lsoa_geometries_gdf: gpd.GeoDataFrame,
+    la_name: str,
+    target_crs: str,
+) -> gpd.GeoDataFrame:
+    """
+    Merges the LA's heat pump suitability scores (pandas DataFrame) with its LSOA geometries (GeoDataFrame).
+    Ensures the final merged result is also a GeoDataFrame in the target CRS.
+
+    Args:
+        hp_suitability_scores_pd (pd.DataFrame): DataFrame containing columns like 'LSOA21CD' or 'DESNZ_pilot_fraction'.
```

Comment:
```suggestion
        hp_suitability_scores_pd (pd.DataFrame): DataFrame containing columns like 'LSOA21CD' and 'DESNZ_pilot_fraction'.
```
is this meant to say 'and'? also, should it be `lsoa` column instead of `LSOA21CD`?

Reply:
Yep, good point